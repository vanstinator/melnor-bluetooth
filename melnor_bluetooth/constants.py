""" Constants for Melnor Bluetooth. """


def _uuid(address: str) -> str:
    return f"0000{address}-0000-1000-8000-00805f9b34fb"


BATTERY_UUID = _uuid("ec08")
DEVICE_USER_NAME_UUID = _uuid("ec01")
UPDATED_AT_UUID = _uuid("ec09")
MODEL_NUMBER_UUID = _uuid("2a24")
MANUFACTURER_UUID = _uuid("2a29")

VALVE_MANUAL_SETTINGS_UUID = _uuid("ec0b")
VALVE_ON_OFF_UUID = _uuid("ec0a")

VALVE_MANUAL_STATES_UUID = _uuid("ec06")

VALVE_0_MODE_UUID = _uuid("ec0f")
VALVE_1_MODE_UUID = _uuid("ec10")
VALVE_2_MODE_UUID = _uuid("ec11")
VALVE_3_MODE_UUID = _uuid("ec12")

# Manufacturers
EDEN = "Eden"
MELNOR = "Melnor"


MODEL_BRAND_MAP = {
    "5901": EDEN,
    "5902": EDEN,
    "5903": EDEN,
    "5904": EDEN,
    "5905": EDEN,
    "5906": EDEN,
    "5907": MELNOR,
    "5908": MELNOR,
    "5909": MELNOR,
    "5910": MELNOR,
    "5911": MELNOR,
    "5912": MELNOR,
    "5913": EDEN,
    "5914": EDEN,
    "5915": EDEN,
    "5916": EDEN,
    "5917": EDEN,
    "5918": EDEN,
}

MODEL_NAME_MAP = {
    "5901": "25443",
    "5902": "25439",
    "5903": "25442",
    "5904": "25438",
    "5905": "25441",
    "5906": "25437",
    "5907": "93281",
    "5908": "93280",
    "5909": "93101",
    "5910": "93100",
    "5911": "93016",
    "5912": "93015",
    "5913": "25443",
    "5914": "25439",
    "5915": "25442",
    "5916": "25438",
    "5917": "25441",
    "5918": "25437",
}

MODEL_SENSOR_MAP = {
    "5901": True,
    "5902": False,
    "5903": True,
    "5904": False,
    "5905": True,
    "5906": False,
    "5907": True,
    "5908": False,
    "5909": True,
    "5910": False,
    "5911": True,
    "5912": False,
    "5913": True,
    "5914": False,
    "5915": True,
    "5916": False,
    "5917": True,
    "5918": False,
}

MODEL_VALVE_MAP = {
    "5901": 4,
    "5902": 4,
    "5903": 2,
    "5904": 2,
    "5905": 1,
    "5906": 1,
    "5907": 4,
    "5908": 4,
    "5909": 2,
    "5910": 2,
    "5911": 1,
    "5912": 1,
    "5913": 4,
    "5914": 4,
    "5915": 2,
    "5916": 2,
    "5917": 1,
    "5918": 1,
}
