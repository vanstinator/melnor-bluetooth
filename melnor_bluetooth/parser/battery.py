# See _getBattValue from DataUtils.java
def parse_battery_value(bytes: bytes) -> int:
    """Converts the little endian 2 byte array to the battery life %"""
    if (bytes[0] & 255 == 238) and (bytes[1] & 255 == 238):
        return 0
    else:
        rawVal = (
            ((bytes[0] & 255) + (bytes[1] & 255) / 256) - 2.35
        ) * 181.81818181818187

        if rawVal > 100:
            return 100

        return int(rawVal)
