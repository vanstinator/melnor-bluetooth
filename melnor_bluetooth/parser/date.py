import datetime
import zoneinfo

from tzlocal import get_localzone


def _time_offset(tz: datetime.tzinfo = get_localzone()):
    """
    Returns the archaic timezone offset in seconds.

    The valve accepts a 4 byte unsigned integer and the official app
    appears to deduct ~30 years from the current time to fit into a 4 byte size.


    All watering operations are keyed off this value and the mobile app _and valves_
    will show bad info we don't replicate the algorithm
    """

    base_time = datetime.datetime.now(tz=zoneinfo.ZoneInfo("Asia/Shanghai"))
    local_time = datetime.datetime.now(tz=tz)

    base_offset = base_time.utcoffset()
    local_offset = local_time.utcoffset()

    if (
        base_offset is None
        or local_offset is None
        or base_time is None
        or local_time is None
    ):
        # TODO log or throw an exception here. caller should handle this
        return 0

    if base_time.dst() is not None:
        base_offset = base_offset - base_time.dst()  # type: ignore

    if local_time.dst() is not None:
        local_offset = local_offset - local_time.dst()  # type: ignore

    if base_offset is None or local_offset is None:
        # TODO log or throw an exception here. caller should handle this
        return 0

    return int(base_offset.total_seconds() - local_offset.total_seconds())


def time_shift(tz: datetime.tzinfo = get_localzone()) -> int:
    return -_time_offset(tz) - 946656000


def get_timestamp(tz: datetime.tzinfo = get_localzone()) -> int:
    """
    Returns the current timestamp as a byte array.
    """

    return int(datetime.datetime.now(tz).timestamp() + time_shift(tz))
