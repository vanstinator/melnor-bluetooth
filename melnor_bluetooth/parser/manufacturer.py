from melnor_bluetooth.constants import MELNOR


def parse_manufacturer(bytes: bytes) -> str:
    """Converts the little endian 2 byte array to the battery life %"""

    string = bytes.decode("utf-8")

    if string.startswith("ML_"):
        return MELNOR

    return string
