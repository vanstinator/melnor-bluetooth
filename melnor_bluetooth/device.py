""" Device interactions for Melnor bluetooth devices. """

from __future__ import annotations

import asyncio
import logging
import struct
from datetime import datetime, time
from typing import List

from bleak.backends.device import BLEDevice
from bleak.exc import BleakError
from bleak_retry_connector import BleakClient  # type: ignore - this is a valid import
from bleak_retry_connector import establish_connection
from deprecated import deprecated

from .constants import (
    BATTERY_UUID,
    MANUFACTURER_UUID,
    UPDATED_AT_UUID,
    VALVE_0_MODE_UUID,
    VALVE_1_MODE_UUID,
    VALVE_2_MODE_UUID,
    VALVE_3_MODE_UUID,
    VALVE_MANUAL_SETTINGS_UUID,
    VALVE_MANUAL_STATES_UUID,
    VALVE_ON_OFF_UUID,
)
from .models.frequency import Frequency
from .utils import date
from .utils.battery import parse_battery_value
from .utils.lock import GLOBAL_BLUETOOTH_LOCK, bluetooth_lock

_LOGGER = logging.getLogger(__name__)


class Valve:
    """Wrapper class to handle interacting with individual valves on a Melnor timer"""

    _device: Device
    _frequency: Frequency
    _id: int
    _is_watering: bool
    _is_frequency_schedule_enabled: bool
    _manual_minutes: int

    def __init__(self, identifier: int, device) -> None:
        self._device = device
        self._frequency = Frequency()
        self._id = identifier
        self._is_watering = False
        self._is_frequency_schedule_enabled = False
        self._manual_minutes = 20
        self._end_time = 0

    def update_state(self, raw_bytes: bytes, uuid: str) -> None:
        """Update the state of the valve from the raw bytes"""

        offset = self._id * 5

        if uuid == VALVE_MANUAL_SETTINGS_UUID:
            # Parses a 5 byte segment from the device and updates the state of the zone
            # [
            #     0   - 0x00, # is_watering - boolean
            #     1-2 - 0x00, # manual_watering_time - unsigned short
            #     3-4 - 0x00, # duplicate of byte 1
            # ]

            self._is_watering = struct.unpack_from(">?", raw_bytes, offset)[0]
            self._manual_minutes = struct.unpack_from(">H", raw_bytes, offset + 1)[0]

        elif uuid == VALVE_MANUAL_STATES_UUID:
            # byte segment for manual watering time left
            # [
            #     0   - 0x00, # unclear, 0-2
            #     1-4 - 0x00, # timestamp - unsigned int
            # ]

            parsed_time = self._end_time = struct.unpack_from(
                ">I", raw_bytes, offset + 1
            )[0]

            self._end_time = parsed_time - date.time_shift() if parsed_time != 0 else 0

        elif uuid == VALVE_ON_OFF_UUID:
            self._is_frequency_schedule_enabled = struct.unpack_from(
                ">?", raw_bytes, self._id
            )[0]

        elif (
            (self._id == 0 and uuid == VALVE_0_MODE_UUID)
            or (self._id == 1 and uuid == VALVE_1_MODE_UUID)
            or (self._id == 2 and uuid == VALVE_2_MODE_UUID)
            or (self._id == 3 and uuid == VALVE_3_MODE_UUID)
        ):
            # byte segment for valve mode
            # [
            #     0   - 0x00, # unclear, always 0
            #     1-4 - 0x00, # timestamp - unsigned int
            #     5-6 - 0x00, # duration - unsigned short
            #     7   - 0x00, # frequency - unsigned char
            # ]

            self._frequency.update_state(raw_bytes)

    @property
    def frequency_bytes(self) -> bytes | None:
        """Returns the frequency bytes"""
        if self._frequency is not None:
            return self._frequency.to_bytes()

    @property
    def id(self) -> int:
        """Returns the raw valve id"""
        return self._id

    @property
    def is_watering(self) -> bool:
        """Returns the zone watering state"""
        return self._is_watering == 1

    @property
    def schedule_enabled(self) -> bool:
        """Returns the zone watering state"""
        return self._is_frequency_schedule_enabled

    @property
    def frequency(self) -> Frequency:
        """Returns the zone watering state"""
        return self._frequency

    @property
    def next_cycle(self) -> datetime | None:
        """Returns the next cycle time"""
        if self.schedule_enabled:
            return self._frequency.next_run_time

        return None

    @is_watering.setter
    @deprecated(version="0.0.18", reason="Use set_is_watering instead")
    def is_watering(self, value: bool) -> None:
        """
        @deprecated
        Sets the zone watering state
        This isn't necessarily safe on the event loop when used in
        conjunction with `Device.push_state`
        if other processes are calling `Device.fetch_state`"""
        self._is_watering = value

    @bluetooth_lock
    async def set_is_watering(self, value: bool) -> None:
        """Atomically sets zone watering state"""
        self._is_watering = value
        await self._device._unsafe_push_state()  # pylint: disable=protected-access

    @property
    def manual_watering_minutes(self) -> int:
        """Returns the number of seconds the zone has been manually watering"""
        return self._manual_minutes

    @manual_watering_minutes.setter
    @deprecated(version="0.0.18", reason="Use set_manual_watering_minutes instead")
    def manual_watering_minutes(self, value: int) -> None:
        """
        @deprecated
        Sets the number of seconds the zone has been manually watering.
        This isn't necessarily safe on the event loop when used in
        conjunction with `Device.push_state`
        if other processes are calling `Device.fetch_state`"""
        self._manual_minutes = value

    @bluetooth_lock
    async def set_manual_watering_minutes(self, value: int) -> None:
        """Atomically set the number of seconds the valve should water."""
        self._manual_minutes = value
        await self._device._unsafe_push_state()  # pylint: disable=protected-access

    @bluetooth_lock
    async def set_frequency_interval_hours(self, value: int) -> None:
        """Atomically set the frequency interval hours"""
        if self._frequency is not None:
            self._frequency.interval_hours = value
            await self._device._unsafe_push_state()  # pylint: disable=protected-access

    @bluetooth_lock
    async def set_frequency_duration_minutes(self, value: int) -> None:
        """Atomically set the frequency duration"""
        if self._frequency is not None:
            self._frequency.duration_minutes = value
            await self._device._unsafe_push_state()  # pylint: disable=protected-access

    @bluetooth_lock
    async def set_frequency_start_time(self, value: time) -> None:
        """Atomically set the frequency start time"""
        if self._frequency is not None:
            self._frequency.start_time = value
            await self._device._unsafe_push_state()  # pylint: disable=protected-access

    @bluetooth_lock
    async def set_frequency_enabled(self, value: bool) -> None:
        """Atomically set the frequency enabled state"""
        self._is_frequency_schedule_enabled = value
        await self._device._unsafe_push_state()  # pylint: disable=protected-access

    @property
    def watering_end_time(self) -> int:
        """Read-only unix timestamp in seconds when watering will end. Pulled directly
        from the device. To influence the value use `set_manual_watering_minutes`"""
        return self._end_time

    def _manual_setting_bytes(self) -> bytes:
        """Returns the 5 byte payload to be written to the device"""

        return struct.pack(
            ">?HH",
            self._is_watering,
            self._manual_minutes,
            self._manual_minutes,
        )

    def __str__(self) -> str:
        return (
            f"      Valve(id={self._id}|"
            + f"is_watering={self._is_watering}|"
            + f"manual_minutes={self._manual_minutes}|"
            + f"seconds_left={self._end_time}"
            + ")"
        )


class Device:
    """A wrapper class to interact with Melnor Bluetooth devices"""

    _battery: int
    _ble_device: BLEDevice
    _brand: str
    _connection: BleakClient
    _connection_lock = asyncio.Lock()
    _is_connected: bool
    _model: str
    _sensor: bool
    _valves: List[Valve]
    _valve_count: int

    def __init__(self, ble_device: BLEDevice) -> None:
        self._battery = 0
        self._ble_device = ble_device
        self._is_connected = False
        self._mac = ble_device.address
        self._valves = []

        # The 1 and 2 valve devices still use 4 valve bytes
        # So we'll instantiate 4 valves to mimic that behavior
        # set of bytes too 🤦‍♂️
        for i in range(4):
            self._valves.append(Valve(i, self))

    async def _read_model(self):
        """Initializes the device"""

        manufacturer_data = await self._connection.read_gatt_char(MANUFACTURER_UUID)

        string = manufacturer_data.decode("utf-8")

        self._model = string[0:5]
        self._valve_count = int(string[6:7])

    def disconnected_callback(self, client):  # pylint: disable=unused-argument
        """Callback for when the device is disconnected"""

        _LOGGER.warning("Disconnected from %s", self._mac)
        self._is_connected = False

    @bluetooth_lock
    async def connect(self, retry_attempts=4) -> None:
        """Connects to the device"""

        if self._is_connected or self._connection_lock.locked():
            return

        async with self._connection_lock:
            try:
                _LOGGER.debug("Connecting to %s", self._mac)

                self._connection = await establish_connection(
                    client_class=BleakClient,
                    device=self._ble_device,
                    name=self._mac,
                    disconnected_callback=self.disconnected_callback,
                    max_attempts=retry_attempts,
                    use_services_cache=True,
                )

                self._is_connected = True

                # Bluez handles certain types of advertisements poorly
                # To work around the missing data we grab it here
                # Callers simply need to connect and it'll be populated
                await self._read_model()

                _LOGGER.debug("Successfully connected to %s", self._mac)

            except BleakError:
                _LOGGER.error("Failed to connect to %s", self._mac)
                self._is_connected = False

    @bluetooth_lock
    async def disconnect(self) -> None:
        """Disconnects the device"""
        await self._connection.disconnect()

    async def fetch_state(self) -> None:
        """Updates the state of the device with the given bytes"""

        if not self._is_connected:
            await self.connect(retry_attempts=1)

        async with GLOBAL_BLUETOOTH_LOCK:
            uuids = [
                BATTERY_UUID,
                VALVE_MANUAL_SETTINGS_UUID,
                VALVE_MANUAL_STATES_UUID,
                VALVE_ON_OFF_UUID,
                VALVE_0_MODE_UUID,
                VALVE_1_MODE_UUID,
                VALVE_2_MODE_UUID,
                VALVE_3_MODE_UUID,
            ]

            try:
                bytes_array: List[bytes] = await asyncio.gather(
                    *[self._read(uuid) for uuid in uuids],
                    return_exceptions=True,
                )

                for i, some_bytes in enumerate(bytes_array):
                    uuid = uuids[i]

                    # This is a little awkward, but it's the only single
                    # attribute we read regularly.
                    if uuid == BATTERY_UUID:
                        self._battery = parse_battery_value(some_bytes)

                    for valve in self._valves:
                        some_bytes = uuids.index(uuid)
                        valve.update_state(bytes_array[some_bytes], uuids[i])

            except BleakError as error:
                # Only throw this error if the device is still connected
                if self._is_connected:
                    raise error

    async def _read(self, uuid: str) -> bytes:
        """Reads the given characteristic from the device"""
        return await self._connection.read_gatt_char(uuid)

    async def _unsafe_push_state(self) -> None:
        """Pushes the new state of the device to the device. WARNING: This
        function runs without an internal lock. Public callers should use `push_state`
        instead"""

        on_off = self._connection.services.get_characteristic(
            VALVE_MANUAL_SETTINGS_UUID
        )

        if on_off is not None:
            await self._connection.write_gatt_char(
                on_off.handle,
                (
                    # pylint: disable=protected-access
                    self._valves[0]._manual_setting_bytes()
                    + self._valves[1]._manual_setting_bytes()
                    + self._valves[2]._manual_setting_bytes()
                    + self._valves[3]._manual_setting_bytes()
                ),
                True,
            )

        await self._connection.write_gatt_char(
            VALVE_ON_OFF_UUID,
            struct.pack(
                ">????",
                # pylint: disable=protected-access
                self._valves[0].schedule_enabled,
                self._valves[1].schedule_enabled,
                self._valves[2].schedule_enabled,
                self._valves[3].schedule_enabled,
            ),
            True,
        )

        zone_1_frequency_bytes = self._valves[0].frequency_bytes
        if zone_1_frequency_bytes is not None:
            await self._connection.write_gatt_char(
                VALVE_0_MODE_UUID, zone_1_frequency_bytes, True
            )

        zone_2_frequency_bytes = self._valves[1].frequency_bytes
        if zone_2_frequency_bytes is not None:
            await self._connection.write_gatt_char(
                VALVE_1_MODE_UUID, zone_2_frequency_bytes, True
            )

        zone_3_frequency_bytes = self._valves[2].frequency_bytes
        if zone_3_frequency_bytes is not None:
            await self._connection.write_gatt_char(
                VALVE_2_MODE_UUID, zone_3_frequency_bytes, True
            )

        zone_4_frequency_bytes = self._valves[3].frequency_bytes
        if zone_4_frequency_bytes is not None:
            await self._connection.write_gatt_char(
                VALVE_3_MODE_UUID, zone_4_frequency_bytes, True
            )

        updated_at = self._connection.services.get_characteristic(UPDATED_AT_UUID)

        if updated_at is not None:
            await self._connection.write_gatt_char(
                updated_at.handle,
                date.get_current_time_bytes(),
                True,
            )

    async def push_state(self) -> None:
        """Pushes the new state of the device to the device"""

        async with GLOBAL_BLUETOOTH_LOCK:
            await self._unsafe_push_state()

    @property
    def battery_level(self) -> int:
        """Returns the battery level of the device"""
        return self._battery

    @property
    def brand(self) -> str:
        """Returns the manufacturer of the device"""
        return self._brand

    @property
    def is_connected(self) -> bool:
        """Returns whether the device is currently connected"""
        return self._is_connected

    @property
    def mac(self) -> str:
        """Returns the MAC address of the device"""
        return self._mac

    @property
    def model(self) -> str:
        """Returns the name of the device"""
        return self._model

    @property
    def name(self) -> str:
        """Returns the name of the device"""
        return f"{self._valve_count} Valve Timer"

    @property
    def rssi(self) -> int:
        """Returns the RSSI of the device"""
        return self._ble_device.rssi

    @property
    def valve_count(self) -> int:
        """Returns the number of valves on the device"""
        return self._valve_count

    @valve_count.setter
    def valve_count(self, value: int) -> None:
        """Sets the number of valves on the device"""
        self._valve_count = value

    @property
    def zone1(self) -> Valve:
        """Returns the first zone on the device"""
        return self._valves[0]

    @property
    def zone2(self) -> Valve | None:
        """Returns the second zone on the device"""
        if self._valve_count > 1:
            return self._valves[1]

    @property
    def zone3(self) -> Valve | None:
        """Returns the third zone on the device"""
        if self._valve_count > 2:
            return self._valves[2]

    @property
    def zone4(self) -> Valve | None:
        """Returns the fourth zone on the device"""
        if self._valve_count > 2:
            return self._valves[3]

    def update_ble_device(self, ble_device: BLEDevice) -> None:
        """Updates the cached BLEDevice for the device"""
        self._ble_device = ble_device

    def __str__(self) -> str:
        string = (
            f"{self.__class__.__name__}(\n    battery={self._battery}\n    valves=(\n"
        )
        for valve in self._valves:
            string += f"{valve}\n"
        return f"{string}    )\n)"

    def __getitem__(self, key: str) -> Valve | None:
        if key == "zone1":
            return self.zone1
        elif key == "zone2":
            return self.zone2
        elif key == "zone3":
            return self.zone3
        elif key == "zone4":
            return self.zone4
