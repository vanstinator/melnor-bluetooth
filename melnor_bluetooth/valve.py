import struct

from bleak import BleakClient

from .constants import BATTERY_CHARACTERISTIC_UUID
from .parser.battery import get_batt_val


class Valve:

    _connection: BleakClient
    _mac: str

    def __init__(self, mac: str) -> None:
        self._mac = mac

    async def connect(self) -> None:
        # try:
        self._connection = BleakClient(self._mac)
        await self._connection.connect()

    # except btle.BTLEDisconnectError:
    #     print("Could not connect to device")
    #     return

    async def disconnect(self) -> None:
        await self._connection.disconnect()

    # async def get_battery_life(self) -> int:
    #     battery_characteristic = self._connection.getCharacteristics(
    #         uuid=BATTERY_CHARACTERISTIC_UUID
    #     )[0]
    # return get_batt_val(battery_characteristic.read())


class Zone:

    _id: int
    _is_watering: bool
    _manual_watering_seconds: int

    def __init__(self, id: int) -> None:
        self._id = id
        self._is_watering = False
        self._manual_watering_seconds = 20 * 60

    def update_state(self, bytes: bytes) -> None:
        """Parses a 5 byte response from the device and updates the state of the zone
        [
            0 - 0x00, # is_watering
            1 - 0x00, # number of max unsigned bytes (0-255) in the run time total
            2 - 0x00, # remainder of the run time total in seconds
            3 - 0x00, # see index 1
            4 - 0x00, # see index 2
        ]

        Get the watering total by multiplying the unpacked int in position 2 and adding the int in position 3
        """

        self._is_watering = bytes[self._id * 5] == 1

        self._manual_watering_seconds = (bytes[(self._id * 5) + 1] * 256) + bytes[
            self._id * 5 + 2
        ]

    @property
    def is_watering(self) -> bool:
        """Returns whether the zone is currently watering"""
        return self._is_watering

    @is_watering.setter
    def is_watering(self, value: bool) -> None:
        """Sets the watering state of the zone"""
        self._is_watering = value

    @property
    def manual_watering_seconds(self) -> int:
        """Returns the number of seconds the zone has been manually watering for"""
        return self._manual_watering_seconds

    @manual_watering_seconds.setter
    def manual_watering_seconds(self, value: int) -> None:
        """Set the number of seconds the zone should manually watering for"""
        self._manual_watering_seconds = value
