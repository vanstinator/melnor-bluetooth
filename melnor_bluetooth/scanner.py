""" Cheap scanner implementation for discovering Melnor Bluetooth devices."""

import asyncio
import logging
import sys
from typing import Callable

from bleak import BleakScanner  # type: ignore - this is a valid import
from bleak.backends.device import BLEDevice
from bleak.backends.scanner import AdvertisementData

_LOGGER = logging.getLogger(__name__)

DeviceCallbackType = Callable[[BLEDevice], None]

lock = asyncio.Lock()


def _callback(
    ble_device: BLEDevice,
    ble_advertisement_data: AdvertisementData,
    callback: DeviceCallbackType,
):
    if ble_advertisement_data.manufacturer_data.get(13) is not None:

        if ble_device.address == "C52127E4-C39B-D0A0-22BD-5837BC84AB9C":
            return

        #  we need to ignore the advertisement data for now
        # https://github.com/vanstinator/melnor-bluetooth/issues/17
        # data = ble_advertisement_data.manufacturer_data[13]
        # model_number = f"{data[0]:02x}{data[1]:02x}"

        _LOGGER.debug("Found device %s: %s", ble_device.name, ble_device.address)
        callback(ble_device)


async def scanner(
    callback: DeviceCallbackType,
    scan_timeout_seconds=60,
):
    """
    Scan for devices.

    :param callback: Callback function.
    :param scan_timeout_seconds: Timeout in seconds. Default 60 seconds
    """

    _LOGGER.debug("Scanning for devices")

    _scanner = BleakScanner()

    def _callback_wrapper(
        ble_device: BLEDevice,
        ble_advertisement_data: AdvertisementData,
    ):
        _callback(ble_device, ble_advertisement_data, callback)

    _scanner.register_detection_callback(_callback_wrapper)

    await _scanner.start()
    if "unittest" not in sys.modules.keys():
        await asyncio.sleep(scan_timeout_seconds)
    await _scanner.stop()
