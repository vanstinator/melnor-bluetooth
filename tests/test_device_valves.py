import struct

from melnor_bluetooth.device import Device, Valve
from melnor_bluetooth.parser.battery import parse_battery_value

zone_byte_payload = struct.pack(
    ">20B",
    # Zone 0
    1,
    1,
    104,
    1,
    104,
    # Zone 2
    0,
    0,
    23,
    0,
    23,
    # Zone 3
    0,
    0,
    42,
    0,
    42,
    # Zone 4
    1,
    0,
    13,
    0,
    13,
)

device = Device("FDBC1347-8D0B-DB0E-8D79-7341E825AC2A")


class TestValveZone:
    def test_zone_0_update_state(self):
        zone = Valve(0, device)

        print(zone_byte_payload)

        zone.update_state(zone_byte_payload)

        assert zone.is_watering == True
        assert zone.manual_watering_minutes == 360

    def test_zone_1_update_state(self):
        zone = Valve(1, device)

        print(zone_byte_payload)

        zone.update_state(zone_byte_payload)

        assert zone.is_watering == False
        assert zone.manual_watering_minutes == 23

    def test_zone_2_update_state(self):
        zone = Valve(2, device)

        print(zone_byte_payload)

        zone.update_state(zone_byte_payload)

        assert zone.is_watering == False
        assert zone.manual_watering_minutes == 42

    def test_zone_3_update_state(self):
        zone = Valve(3, device)

        print(zone_byte_payload)

        zone.update_state(zone_byte_payload)

        assert zone.is_watering == True
        assert zone.manual_watering_minutes == 13

    def test_zone_properties(self):
        zone = Valve(0, device)

        zone.is_watering = True
        zone.manual_watering_minutes = 10

        assert zone.is_watering == True
        assert zone.manual_watering_minutes == 10

    def test_zone_defaults(self):
        zone = Valve(0, device)

        assert zone.is_watering == False
        assert zone.manual_watering_minutes == 20

    def test_zone_byte_payload(self):
        zone = Valve(0, device)

        zone.is_watering = True
        zone.manual_watering_minutes = 10

        assert zone.byte_payload == [1, 0, 10, 0, 10]
