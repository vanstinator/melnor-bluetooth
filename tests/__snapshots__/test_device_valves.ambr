# name: TestDevice.test_str
  '
  Device(
      battery=0
      valves=(
        Valve(id=0|is_watering=False|manual_minutes=20|seconds_left=0)
        Valve(id=1|is_watering=False|manual_minutes=20|seconds_left=0)
        Valve(id=2|is_watering=False|manual_minutes=20|seconds_left=0)
        Valve(id=3|is_watering=False|manual_minutes=20|seconds_left=0)
      )
  )
  '
---
